File: LeetCode/210. 课程表 II.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {number[]}
 */
var findOrder = function (numCourses, prerequisites) {
  const q = [];
  const result = [];
  let neighbors = new Map();
  for (let i = 0; i < numCourses; i++) neighbors.set(i, []);
  let indegree = Array(numCourses).fill(0);
  prerequisites.forEach(([x, pre], i) => {
    neighbors.get(pre).push(x);
    indegree[x]++;
  });
  indegree.forEach((x, i) => x === 0 && q.push(i));
  while (q.length) {
    let cur = q.shift();
    result.push(cur);
    indegree[cur]--;

    for (let neighbor of neighbors.get(cur)) {
      indegree[neighbor]--;
      if (indegree[neighbor] === 0) q.push(neighbor);
    }
  }
  return result.length === numCourses ? result : [];
};

File: LeetCode/210. 课程表 II.rs
impl Solution {
  pub fn find_order(num_courses: i32, prerequisites: Vec<Vec<i32>>) -> Vec<i32> {
      let size:usize=num_courses as usize;
      let mut indegree=vec![0;size];
      let mut neighbors=vec![vec![];size];
      for item in prerequisites.iter(){
          indegree[item[0] as usize]+=1;
          neighbors[item[1] as usize].push(item[0] as usize);
      }
      let mut q:Vec<usize> = (0..num_courses as usize).filter(|&i| indegree[i]==0).collect();
      let mut result=q.clone();
      while !q.is_empty(){
          let mut p=vec![];
          for &i in q.iter(){
              for &neighbor in neighbors[i].iter(){
                  indegree[neighbor]-=1;
                  if indegree[neighbor]==0{
                      p.push(neighbor);
                  }
              }
          }
          result.extend(p.iter());
          q=p;
      }
      match result.len()==size{
          true=>result.iter().map(|&x| x as i32).collect(),
          _=>vec![]
      }
  }
}